File: sample/linked_list.c
#include <stdio.h>
#include <stdlib.h>

/* A linked list node */
typedef struct node {
  int value;
  struct node *next;
} node_t;

int main(void) {
  // Initializing a linked list
  node_t *head = malloc(sizeof(node_t));

  if (head == NULL) {
    exit(1);
  }

  head->value = 1;
  head->next = malloc(sizeof(node_t));

  if (head->next == NULL) {
    exit(1);
  }

  head->next->value = 2;
  head->next->next = NULL;

  // Iterating over a list
  node_t *current = head;

  while (current != NULL) {
    printf("%d\n", current->value);
    current = current->next;
  }

  return 0;
}

File: sample/linked_list.rb
require 'chelsy'

include Chelsy

doc = Document.new

doc.fragments << Directive::Include.new("stdio.h", system: true)
doc.fragments << Directive::Include.new("stdlib.h", system: true)

node_t =
  Typedef.new(:node_t,
    Type::Struct.new(:node, [
      Declaration.new(:value, Type::Int.new),
      Declaration.new(:next, Type::Pointer.new(Type::Struct.new(:node)))]))
node_t.fragments << Comment::Multi.new("A linked list node")
doc << node_t

puts Translator.new(indent_string: '  ').translate(doc)
